### PR TITLE
Harmony 1652 - Batchee/Stitchee support

### PR DESCRIPTION
**Pull request by indiejames**
## Jira Issue ID
HARMONY-1652

## Description
Fixes a problem with jobs not getting marked as complete when the workflow chain has steps that produce more than one output catalog, e.g., Batchee

## Local Test Steps
1. Build the batchee image using the `build_service` script in the repo at https://github.com/danielfromearth/batchee
2. Build the stitchee image using the `build_service` script in the repo at https://github.com/danielfromearth/stitchee
3. Add the follwing to your .env file
4. Add the following to packates/util/env-defaults
```
BATCHEE_SERVICE_QUEUE_URLS='["asdc-trade/batchee:latest,http://localstack:4566/queue/batchee.fifo"]'

STITCHEE_SERVICE_QUEUE_URLS='["asdc-trade/stitchee:latest,http://localstack:4566/queue/stitchee.fifo"]'

BATCHEE_IMAGE=asdc-trade/batchee:latest
BATCHEE_REQUESTS_MEMORY=128Mi
BATCHEE_LIMITS_MEMORY=512Mi
BATCHEE_INVOCATION_ARGS='./docker-entrypoint.sh'
BATCHEE_SERVICE_QUEUE_URLS='["asdc-trade/batchee:latest,http://localstack:4566/queue/batchee.fifo"]'

STITCHEE_IMAGE=asdc-trade/stitchee:latest
STITCHEE_REQUESTS_CPU=128m
STITCHEE_LIMITS_CPU=128m
STITCHEE_REQUESTS_MEMORY=128Mi
STITCHEE_LIMITS_MEMORY=10Gi
STITCHEE_INVOCATION_ARGS='./docker-entrypoint.sh'
STITCHEE_SERVICE_QUEUE_URLS='["asdc-trade/stitchee:latest,http://localstack:4566/queue/stitchee.fifo"]'

```
5. Add the following under the UAT section of services.ymll
```
  - name: batchee-stitchee-concise
    description: |
      TODO - fill in description
    data_operation_version: '0.18.0'
    type:
      <<: *default-turbo-config
      params:
        <<: *default-turbo-params
        env:
          <<: *default-turbo-env
          STAGING_PATH: public/asdc-trade/batchee-stitchee-concise
    umm_s: ''
    capabilities:
      concatenation: true
      concatenate_by_default: true
      extend: true
      default_extend_dimensions: ['mirror_step']
      subsetting:
        bbox: false
        variable: false
        temporal: false
      output_formats:
        - application/netcdf4
      reprojection: false
    steps:
      - image: !Env ${QUERY_CMR_IMAGE}
        is_sequential: true
      - image: !Env ${BATCHEE_IMAGE}
        operations: ['concatenate']
      - image: !Env ${STITCHEE_IMAGE}
        operations: ['extend']
```
7. Run bin/bootstart-harmony followed by bin/start-dev-services (if not running Harmony in a box)
8. Run the following query and verify that it completes using the workflow-ui
```
 curl -Ln -bj  "http://localhost:3000/C1254854453-LARC_CLOUD/ogc-api-coverages/1.0.0/collections/all/coverage/rangeset?concatenate=true&skipPreview=true&maxResults=10"
```
You should see one instance of batchee and two instances of stitchee.

## PR Acceptance Checklist
* [x] Acceptance criteria met
* [x] Tests added/updated (if needed) and passing
* [ ] Documentation updated (if needed)

### CODE REVIEW

File: config/services.yml [hunk 1]
```diff
@@ -87,6 +87,7 @@ https://cmr.earthdata.nasa.gov:
         - application/x-netcdf4
     steps:
       - image: !Env ${QUERY_CMR_IMAGE}
+        is_sequential: true
```
Comment (chris-durbin) [inline config/services.yml:90]: Add example and description of `is_sequential` to the adapting services user guide documentation.

File: services/harmony/app/models/workflow-steps.ts [hunk 2]
```diff
@@ -31,9 +36,16 @@ export interface WorkflowStepRecord {
   // Whether or not this step aggregates the outputs of a previous step
   hasAggregatedOutput: boolean;
 
-  // Whether or no the service should receive a batch of inputs
+  // Whether or not the service should receive a batch of inputs
   isBatched: boolean;
 
+  // Whether or not the service is executed in parallel (the default) or sequentiall, like
```
Comment (chris-durbin) [inline services/harmony/app/models/workflow-steps.ts:42]: sequentially typo

File: services/harmony/app/models/workflow-steps.ts [hunk 3]
```diff
@@ -264,3 +283,54 @@ export async function decrementWorkItemCount(tx: Transaction, jobID, stepIndex):
...
+  const { stepIndex } = step;
+
+  if (step.is_sequential) {
+    const completedCount = await workItemCountForStep(tx, jobID, stepIndex, COMPLETED_WORK_ITEM_STATUSES);
+    const expectedCount =  Math.ceil(numInputGranules / env.cmrMaxPageSize);
+    isComplete = completedCount == expectedCount;
+
+  } else {
+    let prevStepComplete = true;
+    if (stepIndex > 1) {
```
Comment (chris-durbin) [inline services/harmony/app/models/workflow-steps.ts:309]: Seems like there might an issue with this logic if the first workflow step in a chain is not sequential. I don't see any way we could mark the step as completed in this case (probably need to add a test for this scenario - I'm thinking Giovanni might fit this case unless you are marking the giovanni-adapter as sequential).

File: services/harmony/app/models/workflow-steps.ts (hunk 2 above)
Comment (indiejames) [reply services/harmony/app/models/workflow-steps.ts:42]: Fixed

File: config/services.yml (hunk 1 above)
Comment (indiejames) [reply config/services.yml:90]: This will be addressed in HARMONY-1659

File: services/harmony/app/models/workflow-steps.ts (hunk 3 above)
Comment (indiejames) [reply services/harmony/app/models/workflow-steps.ts:309]: This is not an issue

**Review by chris-durbin — APPROVED** [review]
Tested out successfully in sandbox with many different service chains. All requests ended up in terminal states as appropriate.

File: services/harmony/app/models/workflow-steps.ts (hunk 3 above)
Comment (chris-durbin) [reply services/harmony/app/models/workflow-steps.ts:309]: I missed line 308 is marking prevStepComplete = true which prevents the issue I was thinking of.

**Review by ygliuvt — APPROVED** [review]

**Pull request merged**